x-pack/legacy/plugins/apm/public/components/app/Settings/UiIndices/index.tsx

```diff
@@ -0,0 +1,170 @@
+    callApmApi => callApmApi({ pathname: `/api/apm/settings/ui-indices` }),
+    [],
+    { preservePreviousData: false }
+  );
+
+  useEffect(() => {
+    setUiIndices(
+      data.reduce(
+        (acc, { configuration, savedValue }) => ({
+          ...acc,
+          [configuration]: savedValue
+        }),
+        {}
+      )
+    );
+  }, [data]);
+
+  const handleApplyChangesEvent = async (
+    event:
+      | React.FormEvent<HTMLFormElement>
+      | React.MouseEvent<HTMLButtonElement>
+  ) => {
+    event.preventDefault();
+    setIsSaving(true);
+    await saveUiIndices({
+      callApmApi: callApmApiFromHook,
+      uiIndices
+    });
+    setIsSaving(false);
+  };
+
+  const handleChangeIndexConfigurationEvent = async (
+    event: React.ChangeEvent<HTMLInputElement>
+  ) => {
+    const { name, value } = event.target;
+    setUiIndices({
+      ...uiIndices,
+      [name]: value
+    });
+  };
+
+  return (
+    <EuiPanel>
+      <EuiFlexGroup alignItems="center">
+        <EuiFlexItem grow={false}>
+          <EuiTitle>
+            <h2>
+              {i18n.translate('xpack.apm.settings.uiIndicesTitle', {
+                defaultMessage: 'UI Indices'
+              })}
+            </h2>
+          </EuiTitle>
+          <EuiSpacer size="m" />
+          <EuiText size="s" grow={false}>
+            <p>
+              The APM UI uses index patterns to query your APM indices. If
+              you've customized the index names that APM Server writes events
+              to, you may need to update these patterns for the APM UI to work.
+              Settings here take precedence over those set in kibana.yml.
+            </p>
+            <EuiForm>
+              {INDICES.map(({ configuration, label }) => {
+                const matchedConfiguration = data.find(
+                  ({ configuration: config }) => config === configuration
+                );
+                const fallbackValue = matchedConfiguration
+                  ? matchedConfiguration.defaultValue
+                  : '';
+                const savedUiIndexValue = uiIndices[configuration] || '';
+                return (
+                  <EuiFormRow
+                    key={configuration}
+                    label={label}
+                    helpText={`Overrides ${configuration}: ${fallbackValue}`}
+                    fullWidth
+                  >
+                    <EuiFieldText
+                      fullWidth
+                      name={configuration}
+                      placeholder={fallbackValue}
+                      value={savedUiIndexValue}
+                      onChange={handleChangeIndexConfigurationEvent}
+                    />
+                  </EuiFormRow>
+                );
+              })}
+              <EuiSpacer />
+              <EuiFlexGroup justifyContent="flexEnd">
+                <EuiFlexItem grow={false}>
+                  <EuiButton onClick={refetch}>Cancel</EuiButton>
```

Comment:
Why `refetch` on Cancel?

Reply:
cancel behavior resets the values to whatever was previously saved, which can be accomplished by calling `refetch`. Alternatively i could add more to the component state to store the initial values, and set that when cancel is clicked.

Reply:
Ah, makes sense. Seems simpler to just refetch 👍